### PR TITLE
[external metrics] Support HPA v2beta2, v2

### PR DESCRIPTION
**Pull request by juliogreff**
### What does this PR do?

Adds support for v2beta2 and v2 HorizontalPodAutoscalers in the external metrics server, effectively adding support for Kubernetes 1.25+, where support for v2beta1 has been removed.

### Describe how to test/QA your changes

Test that HorizontalPodAutoscaler support works in the external metrics server ([docs](https://docs.datadoghq.com/containers/cluster_agent/external_metrics/?tab=helm)). This needs to be tested in Kubernetes < 1.12 (to use v2beta1), < 1.23 (to use v2beta2), >= 1.23 (to use v2), and >= 1.25 (to ensure that the removed v2beta1 doesn't break).

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by ahmed-mez — COMMENTED** [review]
Let's revert this https://github.com/DataDog/datadog-agent/pull/13340 as part of this PR and make sure the e2e tests pass again.

**Review by ahmed-mez — COMMENTED** [review]
Looks like we have to update the hpa manifests used in e2e tests [here](https://github.com/DataDog/datadog-agent/blob/main/test/e2e/argo-workflows/templates/nginx.yaml#L115) and [here](https://github.com/DataDog/datadog-agent/blob/main/test/e2e/argo-workflows/templates/nginx.yaml#L182)

File: pkg/clusteragent/externalmetrics/autoscaler_watcher.go [hunk 1]
```diff
@@ -76,19 +90,30 @@ func NewAutoscalerWatcher(refreshPeriod int64, autogenEnabled bool, autogenExpir
...
 	// Setup HPA
-	var autoscalerLister autoscaler_lister.HorizontalPodAutoscalerLister
+	var autoscalerLister cache.GenericLister
 	var autoscalerListerSynced cache.InformerSynced
 	if informer != nil {
-		autoscalerLister = informer.Autoscaling().V2beta1().HorizontalPodAutoscalers().Lister()
-		autoscalerListerSynced = informer.Autoscaling().V2beta1().HorizontalPodAutoscalers().Informer().HasSynced
+		hpaGVR, err := autoscalers.DiscoverHPAGroupVersionResource(client)
+		if err != nil {
+			log.Errorf("unable to discover HPA GroupVersionResource: %s", err)
```
Comment (ahmed-mez) [inline pkg/clusteragent/externalmetrics/autoscaler_watcher.go:98]: Shouldn't we fallback to a default version here?

File: pkg/clusteragent/externalmetrics/autoscaler_watcher.go [hunk 2]
```diff
@@ -76,19 +90,30 @@ func NewAutoscalerWatcher(refreshPeriod int64, autogenEnabled bool, autogenExpir
...
+		} else {
+			genericInformerFactory, err := informer.ForResource(hpaGVR)
+			if err != nil {
+				log.Errorf("error creating generic informer: %s", err)
+			}
+
+			autoscalerLister = genericInformerFactory.Lister()
+			autoscalerListerSynced = genericInformerFactory.Informer().HasSynced
```
Comment (ahmed-mez) [inline pkg/clusteragent/externalmetrics/autoscaler_watcher.go:106]: Shouldn't we do this only when `informer.ForResource(hpaGVR)` returns a nil error?

File: pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go [hunk 3]
```diff
@@ -60,6 +66,18 @@ func (f *autoscalerFixture) newAutoscalerWatcher() (*AutoscalerWatcher, kube_inf
 		f.kubeObjects = append(f.kubeObjects, hpa)
 	}
 	kubeClient := kube_fake.NewSimpleClientset(f.kubeObjects...)
+	kubeClient.Resources = []*metav1.APIResourceList{
```
Comment (ahmed-mez) [inline pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go:69]: Can we extend the test suite to cover all the supported api versions?

Comment (juliogreff) [reply pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go:69]: Not sure if unit tests for that would bring a lot of value. I think e2e tests against the different k8s versions are the only way to actually test that support properly.

File: pkg/clusteragent/externalmetrics/autoscaler_watcher.go (hunk 1 above)
Comment (juliogreff) [reply pkg/clusteragent/externalmetrics/autoscaler_watcher.go:98]: We... could? Choosing one is complicated though. `v2`?

File: pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go (hunk 3 above)
Comment (ahmed-mez) [reply pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go:69]: Yes good point. Let's extend `test/e2e/argo-workflows/templates/nginx.yaml` to run multiple tests with different api versions! I know the freeze is approaching so I'll let you decide whether to do it in this PR or later.

File: pkg/clusteragent/externalmetrics/autoscaler_watcher.go (hunk 1 above)
Comment (ahmed-mez) [reply pkg/clusteragent/externalmetrics/autoscaler_watcher.go:98]: I'd say v2 yes but @DataDog/container-ecosystems should be involved in this decision for sure.
The other option is to return the error and not run the external metric provider.

File: pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go (hunk 3 above)
Comment (juliogreff) [reply pkg/clusteragent/externalmetrics/autoscaler_watcher_test.go:69]: I'll create a card to address this not under time pressure :)

File: pkg/clusteragent/externalmetrics/autoscaler_watcher.go (hunk 1 above)
Comment (juliogreff) [reply pkg/clusteragent/externalmetrics/autoscaler_watcher.go:98]: I think returning an error is less surprising behavior, and easier to detect. I'll go with that approach 👍

**Review by ahmed-mez — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

**Review by djmitche — APPROVED** [review]
pkg/util/system/network_stub.go 👍🏽 

Not sure why we own that particular file, but 🤷🏽‍♀️

**Pull request merged**